R/tamper.R

```diff
@@ -0,0 +1,272 @@
+  for (i in rev(seq_len(no_calls))) {
+    calli <- calls[[i]]
+    fname <- calli[[1L]]
+    ## deparse can use more than one line
+    if (!is.na(match(deparse(fname)[1L],
+                     c("methods::.doTrace", ".doTrace")))) {
+      from <- i - 1L
+      break
+    }
+  }
+
+  ## if no trace, look for the first frame from the bottom that is not
+  ## stop, recover or tamper
+  if (from == 0L) {
+    for (i in rev(seq_len(no_calls))) {
+      calli <- calls[[i]]
+      fname <- calli[[1L]]
+      if (!is.name(fname) ||
+          is.na(match(as.character(fname),
+                      c("tamper", "recover", "stop", "Stop")))) {
+        from <- i
+        break
+      }
+    }
+  }
+
+  ## We return here if there are no frames.
+  if (from <= 0L) {
+    cat(gettext("No suitable frames for tamper()\n"))
+    return()
+  }
+
+  if (! interactive()) {
+    try({
+      dump.frames()
+      cat(gettext("tamper called non-interactively; frames dumped, use debugger() to view\n"))
+    })
+    return(NULL)
+
+  } else if (identical(getOption("show.error.messages"), FALSE)) {
+    ## from try(silent=TRUE)?
+    return(NULL)
+  }
+
+  ## ---------------------------------------------------------------
+  ## This is the main stuff
+
+  ## Get the pieces of the pipe, convert to character for printing
+  chr_chain_parts <- chain_parts_to_chr(
+    get("chain_parts", envir = sys.frame(pipe_call))
+  )
+
+  ## Get the bad pipe stage
+  bad_stage <- get_bad_stage(freduce_calls)
+
+  ## Nice printout of pipe stages
+  chr_stages <- format_pipe_stages(chr_chain_parts, bad_stage)
+
+  ## Interactive menu
+  pipe_stack <- TRUE
+  titles <- c("\nEnter a frame number, 1 to switch mode, or 0 to exit  ",
+              "\nEnter a pipe stage number, 1 to switch mode, or 0 to exit  ")
+  other_text <- c("Show pipe stages\n", "Show full stack frames\n")
+
+  stacks <- list(calls, chr_stages)
+
+  repeat {
+
+    which <- menu(
+      c(other_text[pipe_stack + 1], stacks[[pipe_stack + 1]]),
+      title = titles[pipe_stack + 1])
+
+    if (which == 1) {
+      pipe_stack <- ! pipe_stack
+
+    } else if (which == 2) {
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7 - pipe_call)),
+           envir = sys.frame(pipe_call))
+
+    } else if (which > 1 && pipe_stack) {
+      which <- which - 2
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7 - freduce_calls[which])),
+           envir = sys.frame(freduce_calls[which]))
+
+    } else if (which > 1 && ! pipe_stack) {
+      which <- which - 1
+      eval(substitute(browser(skipCalls = skip),
+                      list(skip = 7-which)), envir = sys.frame(which))
+
+    } else {
+      break
+    }
+  }
+
+  ## ---------------------------------------------------------------
+
+}
+
+get_pipe_calls <- function(calls) {
+
+  freduce_calls <- integer()
+  for (i in seq_along(calls)) {
+    calli <- calls[[i]]
+    if (! is.name(calli[[1L]])) next
```

Comment:
If you think so, that's fine with me. Maybe we could check how `freduce` was called, i.e. number and class of arguments, these are always the same in magrittr.